File: src/main/java/ru/mrgrd56/api/translation/TranslationService.java
package ru.mrgrd56.api.translation;

import com.github.benmanes.caffeine.cache.Cache;
import com.github.benmanes.caffeine.cache.Caffeine;
import org.springframework.stereotype.Service;
import ru.mrgrd56.api.exception.NotFoundException;
import ru.mrgrd56.api.translation.translator.Translator;

import java.util.Objects;
import java.util.concurrent.TimeUnit;

@Service
public class TranslationService {
    Cache<TranslationRequestHash, String> cache = Caffeine.newBuilder()
            .maximumSize(3000L)
            .expireAfterWrite(10, TimeUnit.MINUTES)
            .build();

    private final TranslatorProvider translatorProvider;

    public TranslationService(TranslatorProvider translatorProvider) {
        this.translatorProvider = translatorProvider;
    }

    public String translate(String translatorName, String from, String to, String text) {
        Translator translator = translatorProvider.get(translatorName);
        if (translator == null) {
            throw new NotFoundException("Translator not found");
        }

        return cache.get(new TranslationRequestHash(translatorName, from, to, text), k -> {
            return Objects.requireNonNull(translator.translate(text, from, to));
        });
    }

    private record TranslationRequestHash(String translatorName, String from, String to, String text) { }
}

File: src/main/kotlin/ru/mrgrd56/api/translation/TranslationService.kt
package ru.mrgrd56.api.translation

import com.github.benmanes.caffeine.cache.Cache
import com.github.benmanes.caffeine.cache.Caffeine
import org.springframework.stereotype.Service
import java.util.concurrent.TimeUnit

@Service
class TranslationService(private val translatorProvider: TranslatorProvider) {
    private val cache: Cache<TranslationRequestHash, String> = Caffeine.newBuilder()
            .maximumSize(3000L)
            .expireAfterWrite(10, TimeUnit.MINUTES)
            .build()

    fun translate(translatorName: String, from: String, to: String, text: String): String {
        return cache.get(TranslationRequestHash(translatorName, from, to, text)) {
            translatorProvider[translatorName].translate(text, from, to)!!
        }
    }

    private data class TranslationRequestHash(
        val translatorName: String,
        val from: String,
        val to: String,
        val text: String
    )
}
